service/history/timerQueueProcessor.go

```diff
@@ -717,6 +722,42 @@ Update_History_Loop:
 	return ErrMaxAttemptsExceeded
 }
 
+func (t *timerQueueProcessorImpl) processWorkflowTimeout(
+	context *workflowExecutionContext, task *persistence.TimerTaskInfo) error {
+	t.metricsClient.IncCounter(metrics.TimerTaskWorkflowTimeoutScope, metrics.TaskRequests)
+	sw := t.metricsClient.StartTimer(metrics.TimerTaskWorkflowTimeoutScope, metrics.TaskLatency)
+	defer sw.Stop()
+
+Update_History_Loop:
+	for attempt := 0; attempt < conditionalRetryCount; attempt++ {
+		msBuilder, err1 := context.loadWorkflowExecution()
+		if err1 != nil {
+			return err1
+		}
+
+		if !msBuilder.isWorkflowExecutionRunning() {
+			return nil
+		}
+
+		timeoutEvent := msBuilder.AddTimeoutWorkflowEvent()
+		if timeoutEvent == nil {
+			// Unable to add WorkflowTimeout event to history
+			return &workflow.InternalServiceError{Message: "Unable to add WorkflowTimeout event to history."}
```

Comment:
What does returning InternalServiceError buys you here.  This is not something user facing, it will cause the task to get stuck and cause panic for the shard.